src/main.c

```diff
@@ -351,18 +362,120 @@ key_binding_func_key (MBWindowManager   *wm,
   hd_dbus_send_event (s);
 }
 
+/* Close all input devices previously opened */
+static void
+close_input_devices (Display *dpy)
+{
+  int i;
+
+  for (i = 0; i < xi_devices->len; i++)
+    {
+      hd_xi_device *xi_dev = &g_array_index (xi_devices, hd_xi_device, i);
+
+      if (xi_dev->dev)
+        {
+          XCloseDevice (dpy, xi_dev->dev);
+          xi_dev->dev = NULL;
+        }
+    }
+}
+
+/* Fill in the array with input devices type is touchscreen */
+static void
+enumerate_input_devices (Display *dpy)
+{
+  static Atom atom_touchscreen = None;
+  XDeviceInfo *devinfo;
+  int i, ndev;
+  GArray *eclass;
+
+  if (xi_devices)
+    {
+      close_input_devices (dpy);
+      g_array_free (xi_devices, TRUE);
+      xi_devices = NULL;
+    }
+
+  if (atom_touchscreen == None)
+    atom_touchscreen = XInternAtom (dpy, XI_TOUCHSCREEN, True);
+
+  if (atom_touchscreen == None)
+    return;
+
+  /* get XInput DeviceMotion events */
+  devinfo = XListInputDevices (dpy, &ndev);
+
+  eclass = g_array_new (FALSE, FALSE, sizeof (XEventClass));
+  xi_devices = g_array_new (FALSE, TRUE, sizeof (hd_xi_device));
+
+  for (i = 0; i < ndev; i++)
+    {
+      XDeviceInfo info = devinfo[i];
+
+      if (info.use == IsXExtensionPointer)
```

Comment:
Shouldn't it be `if (info.use == IsXPointer || info.use == IsXExtensionPointer)`?

Reply:
I think IsXPointer is for core pointer devices, i.e. those that are not a real mice or TSes. Do you think IsXPointer shall be included and why?

Reply:
Isn't it possible that the user configure the touchscreen as a core pointer?

Reply:
> i.e. those that are not a real mice or TSes.

Could you please share a link to documentation about this?
I wasn't aware of it (core pointer devices to be virtual) and couldn't fine an information.

Reply:
From xorg.conf (5) manpage:

`If hotplugging is disabled, there will normally be at least two: one for the core (primary) keyboard and one for the core pointer. If either of these two is missing, a default configuration for the missing ones will be used. In the absence of an explicitly specified core input device, the first InputDevice marked as CorePointer (or CoreKeyboard) is used.`

Reply:
The point is that we wan't to make a difference whether a device is mouse or ts. Core pointers' type atom is null, so it makes no sense checking them. So, if we have core pointers only, cursor will be hidden initially and that's it.